### PR TITLE
feat: Zoho Calendar

### LINKED ISSUE #5451: [CAL-345] Integration with Zoho calendar
### Is your proposal related to a problem?

I'd like [cal.com](http://cal.com) to allow me to connect my Zoho calendar for work.

### Describe the solution you'd like

I'd like the integration to work between Zoho calendar and [Cal.com](http://Cal.com)

### Describe alternatives you've considered

Tried using a free Google calendar and then sharing that Google calendar with my Zoho calendar for it to appear.  This is so backwards and brings more softwares into the mix then there needs to be.

### Additional context

> From [SyncLinear.com](https://synclinear.com)

### PR DESCRIPTION
**Pull request by murtajaziad**
## What does this PR do?

An app to integrate Zoho Calendar and Cal.com using the official Zoho Calendar API.

Fixes #5451 

**Environment**: Staging(main branch)

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

No automated tests are set up for this. 
You can manually test the integration by following these steps:
1. Link your Zoho account with your Cal.com account.
2. Set your default calendar for creating the events to your Zoho Calendar.
3. Book an event to try the integration.

## Checklist

- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

---

/claim #5451

### CODE REVIEW

**Comment by hariombalhara** [comment]
@murtajaziad Thanks for the PR !! I am wondering if https://github.com/calcom/cal.com/pull/7182 added similar functionality already or not. Also, if it didn't would it make sense to add the functionality into that existing app.

**Comment by murtajaziad** [comment]
> @murtajaziad Thanks for the PR !! I am wondering if #7182 added similar functionality already or not. Also, if it didn't would it make sense to add the functionality into that existing app.

HI @hariombalhara, I have just checked #7182 and it doesn't include any Zoho Calendar functionality, it uses the events/meetings module that's provided by the CRM API. Also, Zoho Calendar and Zoho CRM are two different products that apparently don't overlap.

**Comment by hariombalhara** [comment]
@PeerRich can you confirm if we want to do it in a separate app?

File: packages/app-store/zohocalendar/DESCRIPTION.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+items:
+  - /api/app-store/zohocalendar/ZCal1.jpg
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/DESCRIPTION.md:3]:
```suggestion
  - ZCal1.jpg
```
No need to use full URL. This should also work. It avoids hardcoding app slug and thus simplifies changing it(if needed)

File: packages/app-store/zohocalendar/DESCRIPTION.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+---
+
+Zoho Calendar is an online business calendar that makes scheduling easy for you. You can use it to stay on top of your schedule and also share calendars with your team to keep everyone on the same page.
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/DESCRIPTION.md:6]: @sydwardrae @Jaibles Please review App description

File: packages/app-store/zohocalendar/_metadata.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import _package from "./package.json";
+
+export const metadata = {
+  name: "Zoho Calendar",
+  description: _package.description,
+  installed: true,
+  type: "zoho_calendar",
+  title: "Zoho Calendar",
+  imageSrc: "/api/app-store/zohocalendar/icon.svg",
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:11]:
```suggestion
  imageSrc: "/api/app-store/zohocalendar/icon.svg",
```
It's deprecated and not needed in any new app. `logo` does the same job

File: packages/app-store/zohocalendar/_metadata.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  variant: "calendar",
+  category: "calendar",
+  categories: ["calendar"],
+  logo: "/api/app-store/zohocalendar/icon.svg",
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:15]:
```suggestion
  logo: "icon.svg",
```

Using relative URL keeps it simple

File: packages/app-store/zohocalendar/_metadata.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  publisher: "Cal.com",
+  rating: 5,
+  reviews: 69,
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:18]:
```suggestion
```

File: packages/app-store/zohocalendar/_metadata.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  installed: true,
+  type: "zoho_calendar",
+  title: "Zoho Calendar",
+  imageSrc: "/api/app-store/zohocalendar/icon.svg",
+  variant: "calendar",
+  category: "calendar",
+  categories: ["calendar"],
+  logo: "/api/app-store/zohocalendar/icon.svg",
+  publisher: "Cal.com",
+  rating: 5,
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:17]:
```suggestion
```

File: packages/app-store/zohocalendar/_metadata.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  reviews: 69,
+  slug: "zoho-calendar",
+  trending: false,
+  url: "https://cal.com/",
+  verified: true,
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:22]:
```suggestion
```

File: packages/app-store/zohocalendar/_metadata.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  email: "help@cal.com",
+  dirName: "zohocalendar",
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:24]:
```suggestion
```

File: packages/app-store/zohocalendar/_metadata.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import type { AppMeta } from "@calcom/types/App";
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/_metadata.ts:1]: Replacing this file with config.json. It is deprecated and AppStore CLI doesn't create it.

File: packages/app-store/zohocalendar/README.md
Comment (hariombalhara) [inline packages/app-store/zohocalendar/README.md:1]: @murtajaziad Please add details on how to obtain credentials for Zoho Calendar

File: packages/app-store/zohocalendar/api/add.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    state,
+    prompt: "consent",
+  };
+
+  const query = stringify(params);
+
+  res.status(200).json({ url: `https://accounts.zoho.com/oauth/v2/auth?${query}` });
+}
+
+export default defaultHandler({
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/api/add.ts:38]: Wrapped up endpoints in defaultHandler

**Review by hariombalhara — DISMISSED** [review]
Left some comments

File: packages/app-store/zohocalendar/api/callback.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+import prisma from "@calcom/prisma";
+
+import { decodeOAuthState } from "../../_utils/decodeOAuthState";
+import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import getInstalledAppPath from "../../_utils/getInstalledAppPath";
+import type { ZohoAuthCredentials } from "../types/ZohoCalendar";
+
+const log = logger.getChildLogger({ prefix: [`[[zohocalendar/api/callback]`] });
+
+const zohoKeysSchema = z.object({
```
Comment (hariombalhara) [inline packages/app-store/zohocalendar/api/callback.ts:18]: We can just import it from zod.ts

**Comment by hariombalhara** [comment]
Screenshot 2023-05-10 at 11 54 08 AM
I see `undefined` here.
Also, I think for the same reason I see this error while booking.
Screenshot 2023-05-10 at 12 06 05 PM

**Review by hariombalhara — CHANGES_REQUESTED** [review]
Requesting changes again based on my last comment

**Review by alannnc — CHANGES_REQUESTED** [review]
Made some fixes.
Code looks good, we just missing if createEvent/updateEvent fails throw error so handleNewBooking stops.

**Review by alannnc — APPROVED** [review]
Made some changes and I think it should be good to go now.
Leaving last approve to @hariombalhara

File: packages/app-store/zohocalendar/lib/CalendarService.ts [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+    let eventRespData;
+    const calendarId = externalCalendarId || event.destinationCalendar?.externalId;
+    if (!calendarId) {
+      this.log.error("no calendar id provided in updateEvent");
+      throw new Error("no calendar id provided in updateEvent");
+    }
+    try {
+      const query = stringify({ eventdata: JSON.stringify(this.translateEvent(event)) });
+
+      const eventResponse = await this.fetcher(`/calendars/${calendarId}/events/${eventId}?${query}`, {
```
Comment (joeauyeung) [inline packages/app-store/zohocalendar/lib/CalendarService.ts:168]: This endpoint requires an `etag` that we can get by first retrieving the event on the calendar. This is to ensure that the calendar event wasn't updated while this request is being made.

File: packages/app-store/zohocalendar/lib/CalendarService.ts [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+      throw error;
+    }
+  }
+
+  /**
+   * @param uid
+   * @param event
+   * @returns
+   */
+  async deleteEvent(uid: string, event: CalendarEvent, externalCalendarId?: string) {
```
Comment (joeauyeung) [inline packages/app-store/zohocalendar/lib/CalendarService.ts:199]: Also getting a missing `etag` error for this method.

**Review by joeauyeung — CHANGES_REQUESTED** [review]

**Comment by murtajaziad** [comment]
I have made all the required changes, fyi. auto-generated files are being auto created and added to git by a pre-commit hook.

**Pull request closed without merging**